src/auth-guard/auth-guard.ts

```diff
@@ -0,0 +1,40 @@
+import { Injectable, InjectionToken } from '@angular/core';
+import { CanActivate, ActivatedRouteSnapshot, RouterStateSnapshot, UrlTree, Router } from '@angular/router';
+import { Observable, of, pipe, UnaryFunction } from 'rxjs';
+import { map, switchMap, take } from 'rxjs/operators'
+import { User, auth } from 'firebase/app';
+import { AngularFireAuth } from '@angular/fire/auth';
+
+export const EnableRouterGuardListeners = new InjectionToken<boolean>('angularfire2.enableRouterGuardListeners');
+
+export type AuthPipeGenerator = (next: ActivatedRouteSnapshot, state: RouterStateSnapshot) => AuthPipe;
+export type AuthPipe = UnaryFunction<Observable<User|null>, Observable<boolean|any[]>>;
+
+@Injectable()
+export class AngularFireAuthGuard implements CanActivate {
+
+  constructor(private afAuth: AngularFireAuth, private router: Router) {}
+
+  canActivate(next: ActivatedRouteSnapshot, state: RouterStateSnapshot) {
+    const authPipeFactory: AuthPipeGenerator = next.data.authGuardPipe || (() => loggedIn);
+    return this.afAuth.user.pipe(
+        take(1),
+        authPipeFactory(next, state),
+        map(canActivate => typeof canActivate == "boolean" ? canActivate : this.router.createUrlTree(canActivate))
+    );
+  }
+
+}
+
+export const canActivate = (pipe: AuthPipe|AuthPipeGenerator) => ({
+    canActivate: [ AngularFireAuthGuard ], data: { authGuardPipe: pipe.name === "" ? pipe : () => pipe}
+});
+
+export const loggedIn: AuthPipe = map(user => !!user);
+export const isNotAnonymous: AuthPipe = map(user => !!user && !user.isAnonymous);
+export const idTokenResult = switchMap((user: User|null) => user ? user.getIdTokenResult() : of(null));
```

Comment:
I think `getIdTokenResult` returns a Promise - can Angular handle that?

Reply:
Yeah, switchMap will handle that automatically